### ISSUE TITLE
reporting mistmatch between jtl reports and grafana reports

### ISSUE THREAD

**Issue by yalem-a** [Rbillon59/jmeter-k8s-starterkit#34]
Hello,

Can someone explain why there is deviation between readings on grafana and jtl reports 
**Examples**
According to my test based on my-scenario.jmx which has been shared by @Rbillon59 I see the following differences

1. Total average throughput on grafana is approximately 18 /sec whereas on jtl summary report it is shown as 2.8/sec.
2. throughput by samplers on grafana for http request is 9.43/sec on jtl summary report it is shown as 56.8/min.
3. The same is true of response times as well

Kindly find attached screenshots.
<img width="895" alt="grafana_report_response_time" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/48199750/eafd5ca2-2422-4ed6-935a-0f878ef9a158">
<img width="905" alt="grafana_report_throughput" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/48199750/c108e1a1-979c-4a58-9396-d3c809fa57df">
<img width="710" alt="jtl_report_summary_report" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/48199750/e5e8dd2c-c7bc-4c80-bb82-1e481228c973">

**Comment by Rbillon59** [comment]
Hello,

Thank you for the report.

Actually, your observation is true about the total throughput. The issue is related to the "all" transaction is shown alongside all the samplers, so the total throughput is actually doubled.

I pushed a fix to remove the 'all' transactions to the graph : https://github.com/Rbillon59/jmeter-k8s-starterkit/commit/b530517e1eaa1b3001aba782b776e2eb181b5c80

I also double checked the metrics sent to influx vs the JTL : 

<img width="1132" alt="image" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/17565349/9db28e6d-40f9-4aca-912a-6d7a6b37ae5e">

<img width="712" alt="image" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/17565349/82131031-070f-46a6-ae5b-e2de63117301">
<img width="836" alt="image" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/17565349/9d3e180d-8f45-4947-b184-13f382cdf3c4">

So the throughputs are matching on both

And for the response time 

<img width="1442" alt="image" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/17565349/2d3a3418-a4e0-4c6f-a55e-163d05c1d46b">


<img width="1149" alt="image" src="https://github.com/Rbillon59/jmeter-k8s-starterkit/assets/17565349/ace100fe-07d0-43f4-b44d-9b03481317f7">


Are matching as well.

Also, something to check on your side : 
- Low throughput could lead to less precision, as influxdb backend listener is doing data aggregation while the JTL deals with each requests independently. JTL will alway be the most precise
- While you have requests having redirect, your reports (JTL or Grafana) will show multiple time the same request with a -0 -1 -n suffix and when doing a SUM in any ways, it will add them up.


I did the test deploying a dummy nginx app on my cluster and hitting the / page on it, it might be better to test the setup like that instead of hitting google.com (which has redirects)

Please let me know if you need more infos

**Comment by yalem-a** [comment]
Hello,

Thanks for the fix and feedback. _Throughput_ and _average response time_ on jtl and grafana reports are now aligned. (I will retest this next week with internal apps on premise to avoid redirects as you have pointed out).

I don't know if you have observed the following deviations on response time percentiles. 

- _90%_ 
  JTL - 17ms, grafana- 19ms - **there is slight deviation**
- _95% -
  JTL - 51ms, grafana- 23 ms - **there is a big deviation**
- _99%_ 
  JTL - 30ms, grafana- 124ms -  **there is a big deviation**

**Regarding telegraf operator:**

I am seeing the following error as the test begin to run, I have used your latest updated telgraf-operator manifest. I dont see any readings on grafana. 

Brief overview of my setup

- I have not installed Daemonset as I don't have privilege on the entire cluster
- I have modified the service account to exclude cluster wide requests (i only need to read resources on my namespace)

Is there anything i need to modify to limit the telegraf to view objects only in my namespace .

```
2024/04/20 06:46:26 http: TLS handshake error from 192.168.1.1:58568: remote error: tls: bad certificate
2024/04/20 06:46:26 http: TLS handshake error from 192.168.1.1:28396: remote error: tls: bad certificate
2024/04/20 06:46:26 http: TLS handshake error from 192.168.1.1:43327: remote error: tls: bad certificate
2024/04/20 06:46:27 http: TLS handshake error from 192.168.1.1:43207: remote error: tls: bad certificate
2024/04/20 06:46:27 http: TLS handshake error from 192.168.1.1:11033: remote error: tls: bad certificate
```

**Comment by yalem-a** [comment]
Hello,

Just a quick clarification;

Is there a correlation between the no. of injects set in the script and the expected throughput that we have set using constant throughput timer.

**Comment by Rbillon59** [comment]
> Hello,
> 
> 
> 
> Just a quick clarification;
> 
> 
> 
> Is there a correlation between the no. of injectors set while running the script and the expected throughput that we have set using constant throughput timer. 
> 
> 
> 
> 

Hello

Yes, JMeter replicate each scenario on each injector.

So when you set a scenario doing 1 rqps, 10 injectors means 10 rqps.

**Comment by Rbillon59** [comment]
About the response time derivation, I will take a closer look, I also see on the screenshot I made the sample count is doubled. 

About the telegraf opetator, I actually got the exact same error yesterday after the version bump and that's why I also changed the certificate in the PR.

To still deploy the node agent, you could probably just change the kind to deployment and put it only in the right namespace

**Comment by yalem-a** [comment]
Hello @Rbillon59 

Were able to find time to look into the issue?

**Comment by Rbillon59** [comment]
Hello, I've fixed the table as most as I could in #37 

Actually, the real issue for the mismatch is the aggregation of aggregations.
JTL have all individual requests and when JMeter calculates the metrics aggregations (avg, pct90-95-99 etc..) it takes each individual request into account

Influxdb on it's side receive one metric every 2000 requests. So you have the aggregations per 2000 requests. Then when you want to aggregate all those points into a table or an unique value point, you must aggregate those data again. So at the end, you're mathematically inacurrate.

But, the most important part here is not to have the most precise tool, it's to compare comparable values. So there is not a critical necessity to match JMeter and Influxdb results.

Nonetheless, I will take a bit of time to migrate all the influxdb stack to influxdb 2. Because, with influxdb2, you can leverage the RawBackendListener which sends each requests individually to Influxdb and so give you the ability to be as close as possible to JMeter's results.

I will share my progresses in #25

**Comment by Rbillon59** [comment]
Hi,

I gave a try with influxdb2. So actually, the migration would only be "useful" for JMeter, so I ended up having two influxdb instances. One for the jmeter test results (influxdb2) and one for the rest (influxdb1.6).

First issue, the JMeter Raw InfluxDB Listener only outputs the test results, no other metrics (VU, errors etc..)

Second issue, using that configuration, I was able to get the JMeter test results granularly. One point per sample. So yes, the calculation is more precise, but the flux language is a hell. I spent a bit of time reimplementing a JMeter Dashboard written in Flux, but actually, it's quite hard to get good and presentable results.

Third issue : The flux language is being deprecated on influxdb, I don't really want to spend more time working on a migration that will be an issue .

To conclude : 

I'am not likely to add the influxdb2 + flux support to this project because the issues stated above.

You could take a look at the CSV import to influxdb if you really need granular and precise stats
